stl/inc/ranges

```diff
@@ -6837,6 +6837,682 @@ namespace ranges {
+            requires _Zip::_All_random_access<_IsConst, _ViewTypes...> {
+            auto _Modified_iterator = _Lhs;
+            _Modified_iterator -= _Rhs;
+
+            return _Modified_iterator;
+        }
+
+        _NODISCARD_FRIEND constexpr difference_type operator-(const _Iterator& _Lhs, const _Iterator& _Rhs) noexcept(
+            (noexcept(difference_type{
+                 _RANGES distance(_STD declval<_RANGES iterator_t<_Maybe_const<_IsConst, _ViewTypes>>>(),
+                     _STD declval<_RANGES iterator_t<_Maybe_const<_IsConst, _ViewTypes>>>())})
+                && ...) && noexcept(_STD declval<difference_type>() < _STD declval<difference_type>())
+            && _STD is_nothrow_copy_assignable_v<difference_type>) // strengthened
+            requires(_STD sized_sentinel_for<_RANGES iterator_t<_Maybe_const<_IsConst, _ViewTypes>>,
+                _RANGES iterator_t<_Maybe_const<_IsConst, _ViewTypes>>>&&...) {
+            if constexpr (sizeof...(_ViewTypes) == 1) {
+                return difference_type{_STD get<0>(_Lhs._M_Current) - _STD get<0>(_Rhs._M_Current)};
+            } else {
+                auto _Smallest_difference_value =
+                    difference_type{_STD get<0>(_Lhs._M_Current) - _STD get<0>(_Rhs._M_Current)};
+                auto _Smallest_abs_difference_value =
+                    difference_type{_RANGES distance(_STD get<0>(_Lhs._M_Current), _STD get<0>(_Rhs._M_Current))};
+
+                // Start from index 1, since we already evaluated 0.
+                _Iterator_distance_calculation_impl<1>(
+                    _Lhs, _Rhs, _Smallest_difference_value, _Smallest_abs_difference_value);
+
+                return _Smallest_difference_value;
+            }
+        }
+
+        _NODISCARD_FRIEND constexpr auto iter_move(const _Iterator& _Itr) noexcept(
+            (noexcept(_RANGES iter_move(_STD declval<const _RANGES iterator_t<_Maybe_const<_IsConst, _ViewTypes>>&>()))
+                && ...)
+            && (_STD is_nothrow_move_constructible_v<
+                    _RANGES range_rvalue_reference_t<_Maybe_const<_IsConst, _ViewTypes>>> && ...)) {
+            return _Zip::_Tuple_transform(_RANGES iter_move, _Itr._M_Current);
+        }
+
+        friend constexpr void iter_swap(const _Iterator& _Lhs, const _Iterator& _Rhs) noexcept((
+            noexcept(_RANGES iter_swap(_STD declval<_RANGES iterator_t<_Maybe_const<_IsConst, _ViewTypes>>>(),
+                _STD declval<_RANGES iterator_t<_Maybe_const<_IsConst, _ViewTypes>>>()))
+            && ...)) requires(_STD indirectly_swappable<_RANGES iterator_t<_Maybe_const<_IsConst, _ViewTypes>>>&&...) {
+            _Iterator_swap_impl<0>(_Lhs, _Rhs);
+        }
+
+    private:
+        template <_STD size_t _CurrIndex>
+        _NODISCARD static constexpr bool _Iterator_equality_impl(const _Iterator& _Lhs, const _Iterator& _Rhs) noexcept(

```

Comment:
Reason for using recursion?

Reply:
Recursion in and of itself isn't necessarily a bad thing. Since the recursion base and end case are solved for with `if constexpr` at compile time, the compiler knows exactly how many iterations must be executed, so it can effectively unroll the recursion.

Have you tried inspecting the code generated with this implementation? I've done some (very) basic testing, and when building in Release mode, I'm not seeing any actual recursion taking place in the disassembly when iterating through large sets of views.

With that said, I do understand the sentiment. The functions certainly *could* be implemented without recursion; I just struggle to understand whether or not it would be worth the fold expression madness which would ensue. Can we get any additional opinions on this?

Reply:
Those recursions make the code unnecessarily obscure, inefficient, and verbose, which is a bad thing, at least for the `zip `implementation. I don't see how the fold expression gets crazy in the `zip `implementation, it makes the code easier to understand and tidy.
Have you seen other implementations of `zip_view`? Like libc++?

Reply:
I did not look at other implementations of `zip_view` while working on this one. This is in line with the recommendations for code submissions to the MSVC STL, as using code from other projects might result in license conflicts. (Since libc++ is licensed under the permissive Apache 2.0 license, I realize that this is more of a moot point for your given example, but the idea still stands.)

Honestly, I'm not particularly attached to the current implementation of these recursive functions myself. I'm still unsure of whether or not the recursion would have any major performance implications in practice for the reasons I described earlier, but you do make a valid point concerning the readability of the implementation. I'm going to bite the bullet here and trust your judgment by replacing the recursive functions with ones which use fold expressions and other C++ trickery.

When that's finished, I hope that you and others will be able to comment on which version of the implementations they like better for each function. After all, we wouldn't want code which nobody is happy with getting pushed to the `master` branch, now, would we?

Reply:
I'm in favor of the non-recursive approach you changed to (powered by fold expressions etc.). There are three areas of concern: code clarity, runtime performance, and compiletime throughput. Even if runtime performance is equivalent, it's usually better for compiletime throughput to avoid recursion. Basically, it's "expensive" for the compiler to instantiate functions, and even class definitions, whereas working with alias templates and variable templates as much as possible is cheaper. `tuple_cat()` is an example where avoiding recursion ends up being better for both performance and throughput.